src/transport/smtp/pool/async_impl.rs

```diff
@@ -134,10 +146,36 @@ impl<E: Executor> Pool<E> {
         pool
     }
 
+    pub async fn shutdown(&self) -> Result<(), Error> {
+        if let Some(connections) = self.connections.lock().await.take() {
```

Comment:
This doesn't unlock the mutex until the end of the function, so anyone trying to get a connection will wait for shutdown to finish instead of immediately getting an error.

See the print order in: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=370b2dc1bdaf98b485d2ba591930f7fa

Reply:
Addressed as part of the rebase in 13d7982.